### ISSUE TITLE
light/canopy interpolator initial size

### ISSUE THREAD

**Issue by rafaqz** [traitecoevo/plant#263]
Currently the interpolator is initialised with the size `(1e-6, 1e-6, 17, 16)` in the generic parameter-less constructor. I'm not exactly sure why. But initialising it with `(1e-6, 1e-6, 0, 0)` leads to segfaults during tests.

However these tests want it to be initialised with zero size:

https://github.com/traitecoevo/plant/blob/f5ddd7bd7724ddbc1c77c684d292841dab168a33/tests/testthat/test-scm.R#L279-L290

It would be good to understand the reasoning behind these things before I try to fix it.

**Comment by dfalster** [comment]
The interpolator is a split and has parameters. The initialisation isn't the size but the parameters governing the adaptive refinement.  So `(1e-6, 1e-6, 17, 16)` are `atol, rtol,nbase, max_depth`. You don't want any of those to be 0 :) See [here](https://github.com/traitecoevo/plant/blob/f5ddd7bd7724ddbc1c77c684d292841dab168a33/inst/include/plant/adaptive_interpolator.h#L16). I'd suggest leaving those values as they are. 

The size refers to the number of points in the line, which should be zero upon initialisation, as no data have been added yet. Make sense?




I honestly don't know the ins and out. But do want to point the difference between size and parame

**Comment by rafaqz** [comment]
Thanks that makes sense!

I wonder how it has a length of `33` then, instead of `0`. Something is wack

**Comment by rafaqz** [comment]
Looking into this I just don't understand what the difference is meant to be with this test, where the size is actually 33:

https://github.com/traitecoevo/plant/blob/develop/tests/testthat/test-environment.R#L8-L20

**Comment by dfalster** [comment]
Ok, so the length of 33 may make sense after all. That's because the environment isn't empty, because in the FF16 constructor, the environment spline is constructed:
See here https://github.com/traitecoevo/plant/blob/develop/inst/include/plant/models/ff16_environment.h#L43

Even though the patch is empty, it still makes sense to have an environment to work with:

Here's the x and y values:

```
> e$environment_interpolator$x
 [1] 0.00000 0.03125 0.06250 0.09375 0.12500 0.15625 0.18750 0.21875 0.25000 0.28125 0.31250 0.34375 0.37500 0.40625 0.43750 0.46875 0.50000 0.53125 0.56250 0.59375 0.62500 0.65625 0.68750 0.71875
[25] 0.75000 0.78125 0.81250 0.84375 0.87500 0.90625 0.93750 0.96875 1.00000
> e$environment_interpolator$y
 [1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
```

I'm just trying to see where the 33 number comes from

**Comment by dfalster** [comment]
Ok, 33 comes come 17 +16. 

The spline is created by the Adaptive_interpolator [here](https://github.com/traitecoevo/plant/blob/f5ddd7bd7724ddbc1c77c684d292841dab168a33/inst/include/plant/adaptive_interpolator.h#L61). There is a parameter `nbase`, which is sets the number of steps. This is set in the environment constructor (see first comment in this issue). The spline has length 17, but then after calling refine once another 16 points are added. 

So the length 33 seems fine to me!

**Comment by rafaqz** [comment]
Ok so it seems like previously there was an uninitiated state that no longer occurs after the constructors have changed... Is it needed at all or can I just delete that test? It's the only thing blocking the merge

**Comment by dfalster** [comment]
I'm not sure. What's initialised? 

After constructing a  K93_Environment or FF16_Environment, the spline already initialised, ready to use. So size > 0, likely 33.

**Comment by rafaqz** [comment]
It's just the code in the first message. I'll just delete it if thats ok.

**Comment by dfalster** [comment]
Sorry, I'm still unsure what you're talking about. Can you call me to discuss?

**Comment by dfalster** [comment]
This is no longer a an issue. Addressed as part of #267 #262
